### DISCUSSION TITLE
Use numpy objects instead of pandas for core projection logic

### DISCUSSION THREAD

**Discussion by alexeybaran** [fumitoh/modelx#79] [Ideas]
Use numpy objects instead of pandas for core projection logic. Example of modified __init__.py for CashValue_ME_EX1 attached
![image](https://user-images.githubusercontent.com/52170315/222309106-2b856f7a-fe88-42d7-a71c-1dbc197db8cb.png)

[__init__.txt](https://github.com/fumitoh/modelx/files/10872067/__init__.txt)
(updated __init__.txt)

**Comment by fumitoh** [comment]
@alexeybaran the `__init__.txt` attached above doesn't include anything

**Reply by alexeybaran** [reply]
@fumitoh updated the file

**Reply by fumitoh** [reply]
I added the model with your modification as CashValue_EX4. The speed improves by 20-30% from CashValue_EX1. Not a huge gain, but this is probably because the model is simple. The notbook linked below shows the analysis.  It also shows time comparison for each cells.

https://github.com/lifelib-dev/lifelib/blob/main/lifelib/libraries/savings/savings_example4.ipynb

**Reply by alexeybaran** [reply]
The speed was mostly reduced by the inefficiency of formula I put in `inv_return_table()`.
I also optimized lookup in the `surr_charge_rate(t)` formula. I wasn't sure how it is supposed to work with `has_surr_charge()`, if `surr_charge_id()` isn't nan. I changed `surr_charge_id()` to be always a string. 
Updated version attached. Performance can be optimized further using `get_stacktrace` feature.

[__init__.txt](https://github.com/fumitoh/modelx/files/10890524/__init__.txt)

**Reply by fumitoh** [reply]
CashValue_EX4 and the notebook are updated. Now 50% - 60% reduction in run time from CashValue_EX1.
`has_surr_charge()` is not used anymore in EX4, and now setting   `'nan'` to `surr_charge_id()`  should serve the role of indicating no surrender charge. But this may be error prone, because when `surr_charge_id` is not found, then no surrender charge is assumed automatically. For example, when `surr_charge_id()` is `type_3` but the `type_3` column is missing in `surr_charge_table` by accident.

**Reply by alexeybaran** [reply]
Optimized lookup cells `surr_charge_id()` and `premium_pp(t)`. It should half the time.
[__init__.txt](https://github.com/fumitoh/modelx/files/10891846/__init__.txt)

**Reply by fumitoh** [reply]
This one is the same as the previous one.

**Reply by alexeybaran** [reply]
Sorry, I meant `surr_charge_rate(t)`

```
def surr_charge_rate(t):
    ind_row = np.minimum(duration(t), surr_charge_max_idx())
    return surr_charge_table.values.flat[surr_charge_table_column() + ind_row * len(surr_charge_table.columns)]
```

**Reply by fumitoh** [reply]
Updated EX4 and [the notebook](https://github.com/lifelib-dev/lifelib/blob/main/lifelib/libraries/savings/savings_example4.ipynb) again.
Now EX4 runs in 20% to 30% of the time EX1 takes (70%-80% reduction)!
